### PR TITLE
Unit Testing for the Predictive Modeling

### PR DESCRIPTION
**Pull request by jdovalle10**
# Test Suite Documentation

Below is a description of what each unit test in `tests/test_train_model.py` verifies and how it does so.

---

## `test_xgboost_wrapper_fit_and_predict`

- **Purpose:**  
  Ensure that the `XGBoostWrapper` correctly  
  1. cleans column names,  
  2. encodes categorical features to integers,  
  3. records its `classes_` attribute, and  
  4. delegates `fit()`, `predict()`, and `predict_proba()` calls to the underlying model.  

- **How it works:**  
  1. Creates a small DataFrame with:  
     - a categorical column (`cat`)  
     - a numeric column (`num`)  
     - a column name containing special characters (`weird<col>`)  
  2. Wraps a `DummyModel` inside `XGBoostWrapper` and calls `fit(df, y)`.  
     - Verifies `wrapper.classes_` equals `[0, 1]`.  
     - Asserts the dummy’s `.fit_called_with` DataFrame:  
       - has integer-encoded `cat` values  
       - has sanitized column names (`weird_col_`).  
  3. Calls `wrapper.predict(df)` and checks it returns a NumPy array of zeros.  
  4. Calls `wrapper.predict_proba(df)` and verifies a two-column probability matrix where the second column is all ones.

---

## `test_lgbm_wrapper_fit_and_predict`

- **Purpose:**  
  Validate that the `LGBMWrapper` behaves like the XGBoost wrapper:  
  - encodes cats, sets `classes_`, and delegates calls.

- **How it works:**  
  1. Builds a DataFrame with a single categorical column (`cat`) and a numeric column (`val`).  
  2. Wraps a `DummyModel` in `LGBMWrapper`, calls `fit(df, y)`, and checks:  
     - `wrapper.classes_ == [0,1]`  
     - dummy’s `.fit_called_with` DataFrame has integer codes for the categories.  
  3. Calls `predict(df)` and `predict_proba(df)` to ensure they return valid outputs of the correct shape.

---

## `test_load_data`

- **Purpose:**  
  Confirm that `mp.load_data()` reads train/validation/test sets from parquet files at paths returned by `get_data_paths()`.

- **How it works:**  
  1. Writes three identical small DataFrames to temporary parquet files (`train.parquet`, `validation.parquet`, `test.parquet`).  
  2. Monkey-patches `mp.get_data_paths()` to return these file paths under the `"processed"` key.  
  3. Calls `X_train, y_train, X_val, y_val, X_test, y_test = mp.load_data()`.  
  4. Asserts that:  
     - Each returned object has the correct type (`DataFrame` or `Series`).  
     - The `y_train` series matches the known target values `[0, 1, 0]`.

---

## `test_get_model_variants`

- **Purpose:**  
  Verify that `mp.get_model(name)` returns the appropriate classifier instance (or tuple) for each supported model name, and raises an error for unknown names.

- **How it works:**  
  1. Monkey-patches `mp.get_model_config()` to always return a dummy parameter dict.  
  2. Calls `mp.get_model('xgboost')`, `('lightgbm')`, `('randomforest')`, and verifies the returned objects are instances of `XGBClassifier`, `LGBMClassifier`, and `RandomForestClassifier`, respectively.  
  3. Calls `mp.get_model('catboost')` and checks it returns a `(CatBoostClassifier, cat_features)` tuple.  
  4. Ensures `mp.get_model('unknown_model')` raises `ValueError`.

---

## `test_save_and_load_model`

- **Purpose:**  
  Ensure that `mp.save_model(model, name, tuned)` writes the model to disk under the path from `get_paths()['models']`, and that the saved file can be reloaded via `joblib.load()`.

- **How it works:**  
  1. Monkey-patches `mp.get_paths()` to return a temporary folder for `'models'`.  
  2. Calls `mp.save_model(DummyModel(), 'testmodel', tuned=False)` and captures the returned file path.  
  3. Asserts that the file exists at that path.  
  4. Uses `joblib.load()` to reload the file and verifies the loaded object is a `DummyModel`.

---

## `test_plot_feature_importance_and_confusion`

- **Purpose:**  
  Check that plotting utilities generate and save their figure files under `get_paths()['figures']`.

- **How it works:**  
  1. Monkey-patches `mp.get_paths()` to return a temporary folder for `'figures'`.  
  2. **Feature Importance:**  
     - Defines a dummy model (`FIModel`) with a `feature_importances_` array.  
     - Calls `mp.plot_feature_importance(model, feature_names, 'mod')` and verifies the returned path points to an existing file.  
  3. **Confusion Matrix:**  
     - Trains a real `DecisionTreeClassifier` on a small synthetic dataset.  
     - Calls `mp.plot_confusion_matrix(dt, X, y, 'dtmod')` and asserts that file exists.

---

## `test_evaluate_model_on_test`

- **Purpose:**  
  Ensure that `mp.evaluate_model_on_test(model, X, y, name)` returns a metrics dictionary containing all expected keys.

- **How it works:**  
  1. Trains a simple `DecisionTreeClassifier` on a tiny dataset.  
  2. Calls `mp.evaluate_model_on_test(dt, X, y, 'dtmod')`.  
  3. Asserts that the returned dictionary has the keys:  
     ```
     test_accuracy
     test_precision
     test_recall
     test_f1
     test_roc_auc
     ```

---

## `test_save_results`

- **Purpose:**  
  Validate that `mp.save_results(model_name, val_metrics, test_metrics, tuned)` writes results to a CSV under the `results` path and appends on subsequent calls.

- **How it works:**  
  1. Reloads `mp` to clear any prior patches, then monkey-patches `mp.get_paths()` to return a temporary folder for `'results'`.  
  2. Defines dummy metrics dictionaries (`val_metrics`, `test_metrics`).  
  3. **First save:** Calls `mp.save_results(..., tuned=False)`, asserts the CSV file exists, and checks the CSV contains the `val_` and `test_` metric columns.  
  4. **Second save:** Calls `mp.save_results()` again and asserts the CSV now has two rows, confirming appending.

---

### CODE REVIEW

**Comment by arturomedinadb** [comment]
.pyc conflict untrack the file
